Feature: Polynomials and rational functions

lounres opened the pull request:
This is implementation of different kinds of polynomials and rational functions.
It should close issues #462, #336
See also: #39, #446, 

## Comment: files systematisation

1. Each kind of polynomials and corresponding polynomial spaces are placed in corresponding file named after type of polynomials.
2. The same goes for rational functions: each kind of rational functions and corresponding rational functions' spaces are placed in corresponding file named after type of polynomials.
3. Utilities for each kind of polynomials and rational functions are placed in file named after corresponding type of polynomials/rational functions with additional `Util` suffix.
4. Constructors (and building DSLs) for each kind of polynomials and rational functions are placed in file named after corresponding type with additional `Constructors` suffix.
5. Abstract entities are places in `Polynomial.kt` and `RationalFunction.kt`.

## Problems and questions

1. **Is it worth to change `Map<Symbol, UInt>` to [`android.util.ArrayMap<Symbo, UInt>`](https://developer.android.com/reference/kotlin/android/util/ArrayMap) or something similar?**
    This class is designed to be more memory efficient for small numbers of entries which is exactly our use case.

1. **What is better way to implement interoperability with integers?**
    As fact, additive (abelian) groups has natural product operation with integers: product of element `c` and integer `n` is just:
    - sum of `n` copies of `c` if `n` is positive,
    - sum of `-n` copies of `-c` if `n` is negative (or negate of sum of `-n` copies of `c`, which is the same),
    - zero (neutral element) of the group if `n` is zero.
    
    The same goes for multiplicative groups and power. (As a matter of fact, rings do not provide multiplicative (abelian) groups, but multiplicative (abelian) monoids, so the power can be defined not for integers, but for non-negative integers.) In context of rings we also have addition of ring members and integers: for real, sum of element `c` and integer `n` is just `c + n * one` (where `one` is unit of the ring). And multiplication and exponentiation are very helpful and frequently used.

    For example, when you calculate derivative of polynomial you multiply coefficients of terms to corresponding degrees (which are positive integers), so to implement derivative you have to implement multiplication yourself (see `algebraicStub.kt`) or require ability to convert integer to ring member (see `Polynomial<C>.derivative` that preserves former logic). Both variants are bad: manual multiplication usually slower then multiplication that could be provided by corresponding ring, and requiring convertibility (by `ScaleOperations` or even `NumericAlgebra` because there are no alternatives) makes it impossible to use it in common rings (say hi to rational numbers again).

    In my opinion, the best way is to declare the interactions in `GroupOps`/`Group` (for multiplication) and `Ring` (for exponentiation that is already declared and addition) interfaces and implement them in all current algebraic structures. (It's very easy, even is a bit of boilerplate.) It's even inconsistent in some way that exponentiation is declared in `Ring`, but multiplication by `Int`/`UInt` (which is the same operation not in multiplicative group/monoid, but in additive group/monoid) is not.
    
    So the trivial question here: what should I do here?
1. **Maybe, it is better to move first argument `ring` of utility functions (`substitute`, `derivative`, etc.) to context (when context receivers will be available), isn't it?**
1. **Is there need in other functionality?**
    For example, string representation: it is already implemented in my other repositories, it remains only to copy it with minimal modification.
1. **Should RFs spaces be implemented as field?**
    If yes, then what division operations should be implemented? (For example, what should `c / c` return? And when field over `C` is provided?) And how to implement current `Number`-oriented functions of `Field`?
1. **Should cases for `A` (the type of provided ring of constants) being `Field` be considered as well?**
    I mean that function `operator fun Polynomial<C>.div(other: C): Polynomial<C>` makes sense only in context of of field over `C`. And it looks helpful. So should I implement something like this? (But of course, I'll be able to implement something like this when context receivers will be released. For now I can sketch it, comment it and add TODO mark to it.)

## TODO list

- [ ] Write optimized algorithms for substitution.
- [ ] Anything else? (Look at inline TODOs.)

CommanderTvis commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/polynomialUtil.kt:
```diff
@@ -0,0 +1,185 @@
+import kotlin.math.pow
+
+
+// region Utilities
```
We do not usually use regions.

CommanderTvis commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/AbstractPolynomial.kt:
```diff
@@ -0,0 +1,416 @@
+/**
+ * Abstraction of polynomials.
+ */
+public interface AbstractPolynomial<C>
```
Consider removing prefix `Abstract`.

CommanderTvis commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt:
```diff
@@ -5,128 +5,484 @@
  * @param coefficients constant is the leftmost coefficient.
  */
-public class Polynomial<out T>(public val coefficients: List<T>) {
+public data class Polynomial<C>(
```
And this one can be renamed to `ListPolynomial`.

CommanderTvis commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/polynomialUtil.kt:
```diff
@@ -0,0 +1,185 @@
+//    if (degree > -1) Polynomial(coefficients.subList(0, degree + 1)) else zero
+
+/**
+ * Crates a [PolynomialSpace] over received ring.
```
```suggestion
 * Creates a [PolynomialSpace] over a received ring.
```

CommanderTvis commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/polynomialUtil.kt:
```diff
@@ -0,0 +1,185 @@
+    PolynomialSpace(this)
+
+/**
+ * Crates a [PolynomialSpace]'s scope over received ring.
```
```suggestion
 * Creates a [PolynomialSpace]'s scope over a received ring.
```

CommanderTvis commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/NumberedPolynomial.kt:
```diff
@@ -0,0 +1,627 @@
+     * It is recommended not to put zero monomials into the map, but is not prohibited. Lists of degrees always do not
+     * contain any zeros on end, but can contain zeros on start or anywhere in middle.
+     */
+    public val coefficients: Map<List<UInt>, C>
```
`UInt` is second-class citizen in Kotlin, so using this constructor is quite inconvenient. By the way, `Polynomial(List)` is indexed with `Int`, so these classes are incompatible.

lounres replied:
> By the way, `Polynomial(List)` is indexed with `Int`, so these classes are incompatible.

I don't see any reason why I should use this values as indices (of list). They describe powers of variables in each term. So it will be even more headache to check if all this values are non-negative.

Also during writing code for derivatives and antiderivatives (see [here](https://github.com/lounres/kmath/blob/2082175af5e9e811d3a6dfa3079a7dcb2d25a2fe/kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/numberedPolynomialUtil.kt#L341)) for `NumberedPolynomial` I understood that it's better to use `Int` for variables' IDs and `UInt` for their powers and orders of derivative with respect to them. Because, as we already said, `Int` is used only during indexation on that `List`s, and `UInt` is used only for evaluation powers of terms. So it will be unpleasant to use anything else.

lounres replied:
> `UInt` is second-class citizen in Kotlin, so using this constructor is quite inconvenient.

Yeah. This issue is against issue of checking non-negativity of values (previously described issue). Is it really that worse? (Also, will that code really be used somewhere outside of Kotlin assuming that there are already a few classes and functions requiring `UInt`s?)

lounres replied:
Also this constructor is internal, because I don't want user to construct polynomial with dirty term signature representations (see problem 1 for details). So this constructor can't be inconvenient when it is not accessible.

CommanderTvis replied:
> So it will be unpleasant to use anything else.
OK, I agree.

lounres commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt:
```diff
@@ -5,128 +5,484 @@
  */
-public fun Polynomial<Double>.value(arg: Double): Double = coefficients.reduceIndexed { index, acc, c ->
-    acc + c * arg.pow(index)
+internal class PolynomialError : Error {
```
Thank you for the comment. But there are 2 moments:
1. Actually, I did not use it, so I decided just to remove it.
2. There is no documentation for `kotlin.Error`. Only in K/JVM implementation it is a type alias for `java.lang.Error` that is described as (I look in [Java 11 API](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Error.html)):
    > An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions. The ThreadDeath error, though a "normal" condition, is also a subclass of Error because most applications should not try to catch it.
    
    And I was going to use it in cases of strange internal behaviour of polynomials (like in case when you have more complicated structure inside `Polynomial`, and it was used/changed incorrectly, so it looks like addition of two polynomials suddenly crashes). Such cases meet the definition, so `Error` is right choice. Or do I miss something?

CommanderTvis replied:
> , and it was used/changed incorrectly

The convention is that `Error` is used only for abnormal situations related to VM (incorrect bytecode, out-of-memory). Passing incorrect arguments to a function is clearly a `RuntimeException`.

lounres replied:
> The convention is that `Error` is used only for abnormal situations related to VM (incorrect bytecode, out-of-memory).

OK. Got it.

> Passing incorrect arguments to a function is clearly a `RuntimeException`.

I was talking not about case of passing incorrect arguments, but about case when some "system" made mistake that it supposedly should not make.

lounres commented:
So now the main feature is finally ready. It only remains to add tests and write optimized algorithms for substitutions (with tests and benchmarks).

I also have questions about some details which I formulated in "Problems and questions" section in the pull request description message.

altavir commented:
This is a large PR, so review will take some time. In order for this to be accepted, you need to add some examples in the example directory making it obvious how to use it. Tests are also necessary. I think that we will require a design note (in `docs` directory) on all new feature PRs in order to make sure different features play well together. You can probably use your very thorough (👍 ) PR text for that.

lounres commented:
OK. I am going to be busy for now. So I think I'll write examples and design notes some time later (about 1 to 2 weeks later). And after that will deal with tests and rest stuff.

lounres commented:
I added examples and design notes. (Although, I am not sure if they came out good.) Also decided to move optimizations to new refactoring branch (because I found out that there a lot of algorithms to refactor).

Now I am going to gradually add tests...

altavir commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/LabeledPolynomial.kt:
```diff
@@ -0,0 +1,619 @@
+ *
+ * @param C the type of constants.
+ */
+public data class LabeledPolynomial<C>
+@PublishedApi
```
I don't understand what this thing does. Could you provide a reference to its description and usage?

lounres replied:
Sorry, are you asking about [`@PublishedAPI`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-published-api/)?

altavir commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt:
```diff
@@ -5,128 +5,498 @@
+     */
+    @JvmName("unaryPlusConstant")
+    @JsName("unaryPlusConstant")
+    public operator fun C.unaryPlus(): C = this
```
I would like to avoid duplicating algebra functionality this way. It is better to use nested algebra scopes.

altavir commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt:
```diff
@@ -5,128 +5,498 @@
-    acc + c * arg.pow(index)
-}
+@Suppress("INAPPLICABLE_JVM_NAME", "PARAMETER_NAME_CHANGED_ON_OVERRIDE") // FIXME: Waiting for KT-31420
+public interface PolynomialSpace<C, P: Polynomial<C>> : Ring<P> {
```
This class seems to be a needless abstraction. I do not see, where all of its functions are used.

lounres replied:
For me historically it was a systematisation of API that I want to see in polynomial spaces. I also thought there could be cases of need in common abstraction (for some extension functions or something other). But currently there is no actual use case except for systematisation of the API.

lounres commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/NumberedPolynomial.kt:
```diff
@@ -0,0 +1,427 @@
+ */
+public data class NumberedPolynomial<C>
+@PublishedApi
+internal constructor(
+    /**
```
Actually it depends on what level of safety you want to provide. There is a problem in representation of polynomial that type system does not cover: signature of monomial can have extra zeros at the end, which allows (from type system's POV) to declare two different monomials with the same signature. For example, `5x_0 = 2x_0 + 3x_0` can be represented as
```kotlin
mapOf(
    listOf(1u) to 5,
)
```
as well as
```kotlin
mapOf(
    listOf(1u) to 2,
    listOf(1u, 0u) to 3,
)
```
So I see only such solutions of the problem:
1. Ignore it. ("It's not a bug, it's a feature.")
2. Prohibit constructing incorrect numbered polynomials by controlling the way they are constructed.
3. Allow constructing incorrect numbered polynomials and fix them in arithmetic operations.

I assumed that way # 2 is the best. So I closed the constructor and made:
1. fabric that checks and fixes received coefficients map,
2. and functions (see `NumberedPolynomialAsIs` and `NumberedPolynomialWithoutCheck`) that make the constructor implicitly public but warn that there could be problems in case of their misuse.

lounres commented on examples/src/main/kotlin/space/kscience/kmath/functions/polynomials.kt:
```diff
@@ -0,0 +1,397 @@
+fun listPolynomialsExample() {
+    // [ListPolynomial] is a representation of a univariate polynomial as a list of coefficients from the least term to
+    // the greatest term. For example,
+    val polynomial1: ListPolynomial<Int> = ListPolynomial(listOf(2, -3, 1))
```
Yes, I already implemented better API as `ListPolynomial(2, -3, 1)`. You can see it in example 4 string below or its definition on the top of file `kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/listConstructors.kt`.

lounres commented on kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt:
```diff
@@ -5,128 +5,498 @@
  */
-@Suppress("FunctionName")
-public fun <T> Polynomial(vararg coefficients: T): Polynomial<T> = Polynomial(coefficients.toList())
+public interface Polynomial<C>
```
> This abstraction is useless.

Yeah, I agree that this interface is useless and should be removed (or replaced).

> It should at least provide a way to extract individual coefficients. Maybe expose coefficients as `Buffer`

In general multivariate case I don't see a way to expose them, because in that case we should either declare way to represent term with both its signature and its coefficient or declare common order on monomials (independently on number of variables!).

It is possible in univariate case by just representing the polynomial as a sequence. But I don't know if it is necessary.

altavir commented on settings.gradle.kts:
```diff
@@ -26,6 +26,9 @@ include(
     ":kmath-core",
     ":kmath-coroutines",
     ":kmath-functions",
+    ":test-utils-functions",
```
Are those tests used more than in one module? If not, please move them to appropriate module test source sets instead of separate modules.

lounres replied:
Done.

altavir reviewed (changes requested):
Please review necessity of test modules. Otherwise, we will merge the module and then refine the API post-merge

lounres commented:
Done.

altavir commented:
@CommanderTvis waiting for your review

pklimai reviewed (changes requested):
My first comment would be - it's a huge module without README.md in the directory...

altavir commented:
Seems to be failing on Native. @lounres, have you done the full build locally?

altavir commented:
Maybe better to remove native plugin from polynomials for now. We can always add it later.

lounres commented:
Just clean rebuilt it (`:clean` + `:build` as-is and with kover disabled). Everything has been built successfully (and without any problems nor warnings in `kmath-functions` and `kmath-polynomial`). You can logs [here](https://gist.github.com/lounres/d84596524a1b10c5e887b50326065e9c). It's Windows 10 x64.

By the way, a lot of on-push GitHub builds failed. Most of them failed due to bug in kover (seems like [this issue](https://github.com/Kotlin/kotlinx-kover/issues/183)) and some of them failed due to disappearance of daemon.

altavir commented:
It seems like it requires to much resources to run in the github container. I will try to fix it post-merge. Waiting for @CommanderTvis review now.

CommanderTvis reviewed (changes requested):
The first problem is that `Polynomial` interface is duplicated in two modules:
kmath-functions: https://github.com/lounres/kmath/blob/feature/polynomials/kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt#L22
kmath-polynomial: https://github.com/lounres/kmath/blob/feature/polynomials/kmath-polynomial/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt#L17

CommanderTvis commented:
Another doubt is about the type parameter in `interface Polynomial` in kmath-polynomial. It can be just removed without any changes in semantics.

CommanderTvis commented:
Also, about type parameters, many classes in this PR can have `out` or `in` declaration site variances e.g., ListPolynomial<***out*** C>.

CommanderTvis commented:
Some implementations of operators in the contexts can be shortened:
```kotlin
    public override operator fun C.plus(other: Int): C = ring { addMultipliedByDoubling(this@plus, one, other) }
    public override operator fun C.minus(other: Int): C = ring { addMultipliedByDoubling(this@minus, one, -other) }
```

```kotlin

    public override operator fun C.plus(other: Int): C = ring { addMultipliedByDoubling(this@plus, one, other) }
    public override operator fun C.minus(other: Int): C =  this@minus + (-other)
```

lounres commented:
> Another doubt is about the type parameter in `interface Polynomial` in kmath-polynomial. It can be just removed without any changes in semantics.

OK. Will be removed.

lounres commented:
> Also, about type parameters, many classes in this PR can have `out` or `in` declaration site variances e.g., ListPolynomial<_**out**_ C>.

If "many" means "just `ListPolynomial`, `NumberedPolynomial`, `LabeledPolynomial`", then I agree. Because I don't see any other possible variance cases.

lounres commented:
> The first problem is that `Polynomial` interface is duplicated in two modules: kmath-functions: https://github.com/lounres/kmath/blob/feature/polynomials/kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt#L22 kmath-polynomial: https://github.com/lounres/kmath/blob/feature/polynomials/kmath-polynomial/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt#L17

Watching discussion in telegram. Waiting for a final verdict.

lounres commented:
> Some implementations of operators in the contexts can be shortened:
> 
> ```kotlin
>     public override operator fun C.plus(other: Int): C = ring { addMultipliedByDoubling(this@plus, one, other) }
>     public override operator fun C.minus(other: Int): C = ring { addMultipliedByDoubling(this@minus, one, -other) }
> ```
> 
> ```kotlin
>     public override operator fun C.plus(other: Int): C = ring { addMultipliedByDoubling(this@plus, one, other) }
>     public override operator fun C.minus(other: Int): C =  this@minus + (-other)
> ```

Yes, it's true. Although, I don't like it in some way because I don't fully control the `minus` operation's logic but delegate it to another operation. (It was driven by principal that I should control number of used operations because there can be algebraic contexts with very slow operations, so `a + (-b)` would be twice slower than `a - b`. Fortunately, it's not the case in the example.)

But, yeah, I'll fix it.

CommanderTvis commented:
> > Also, about type parameters, many classes in this PR can have `out` or `in` declaration site variances e.g., ListPolynomial<_**out**_ C>.
> 
> If "many" means "just `ListPolynomial`, `NumberedPolynomial`, `LabeledPolynomial`", then I agree. Because I don't see any other possible variance cases. This three will be provided with variance.

Yes, actually I meant them.

lounres commented:
I removed `Polynomial` and fixed variance (and everything seem OK there). But I am currently struggling with code "shortening" (see neighbour branch for current progress). I suspect [KT-31420](https://youtrack.jetbrains.com/issue/KT-31420). So it seems to me that some changes should be applied a bit later (when the issue will be fixed, which is planned on 1.8.0).
